File: src/main/java/com/dyingbleed/akita/source/impl/CanalSource.java
package com.dyingbleed.akita.source.impl;

import com.alibaba.otter.canal.client.CanalConnector;
import com.alibaba.otter.canal.client.CanalConnectors;
import com.alibaba.otter.canal.protocol.CanalEntry.*;
import com.alibaba.otter.canal.protocol.Message;
import com.alibaba.otter.canal.protocol.exception.CanalClientException;
import com.dyingbleed.akita.source.AkitaSource;
import com.dyingbleed.akita.source.AkitaSourceCallback;
import com.dyingbleed.akita.utils.EntryUtils;
import com.google.common.collect.Lists;
import com.google.common.net.HostAndPort;
import org.apache.commons.lang.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetSocketAddress;
import java.util.List;

/**
 * Created by 李震 on 2017/12/4.
 */
public class CanalSource implements AkitaSource {

    /*
     * 常量
     * */

    private static final String SERVER_SEPERATOR = ",";

    // 连续异常重启次数
    private static final int RETRY_TIMES = 5;

    private static final Logger logger = LoggerFactory.getLogger(CanalSource.class);


    /*
     * Canal 参数
     * */

    private String canalServers;

    private String canalDestination;

    private String canalUsername;

    private String canalPassword;

    private String canalFilter;

    /*
     * Canal 连接
     * */

    private CanalConnector canalConnector;


    public CanalSource(String canalServers, String canalDestination, String canalUsername, String canalPassword, String canalFilter) {
        this.canalServers = canalServers;
        this.canalDestination = canalDestination;
        this.canalUsername = canalUsername;
        this.canalPassword = canalPassword;
        this.canalFilter = canalFilter;
    }

    @Override
    public void init() {
        CanalConnector canalConnector;

        if (this.canalServers.contains(SERVER_SEPERATOR)) {
            // 集群模式
            logger.info("使用 Canal Client 集群模式");
            List<InetSocketAddress> hosts = Lists.newLinkedList();
            for (String server: StringUtils.split(this.canalServers, SERVER_SEPERATOR)) {
                HostAndPort hostAndPort = HostAndPort.fromString(server);
                hosts.add(new InetSocketAddress(hostAndPort.getHostText(), hostAndPort.getPortOrDefault(11111)));
            }
            canalConnector = CanalConnectors.newClusterConnector(
                    hosts,
                    canalDestination,
                    canalUsername,
                    canalPassword
            );
            logger.info("创建 Canal 连接：{}, {}, {}, {}", this.canalServers, this.canalDestination, this.canalUsername, this.canalPassword);
        } else {
            // 单机模式
            logger.info("使用 Canal Client 单机模式");
            HostAndPort hostAndPort = HostAndPort.fromString(this.canalServers);
            canalConnector = CanalConnectors.newSingleConnector(
                    new InetSocketAddress(hostAndPort.getHostText(), hostAndPort.getPortOrDefault(11111)),
                    canalDestination,
                    canalUsername,
                    canalPassword
            );
            logger.info("创建 Canal 连接：{}, {}, {}, {}", this.canalServers, this.canalDestination, this.canalUsername, this.canalPassword);
        }

        logger.info("开始连接 Canal Server");
        canalConnector.connect();
        logger.info("订阅 Canal Server {}", canalFilter);
        canalConnector.subscribe(canalFilter);
        logger.info("回滚 Canal Server");
        canalConnector.rollback();

        this.canalConnector = canalConnector;
    }

    @Override
    public void pull(int i, AkitaSourceCallback callback) {
        long messageId = -1;

        try {
            Message message = canalConnector.getWithoutAck(i);
            messageId = message.getId();

            if (messageId != -1 && message.getEntries().size() > 0) {
                logger.info("收到 Canal Server 消息 {}", messageId);
                for (Entry entry: message.getEntries()) {
                    if (entry.getEntryType() == EntryType.TRANSACTIONBEGIN || entry.getEntryType() == EntryType.TRANSACTIONEND) continue;

                    String key = entry.getHeader().getTableName();
                    try {
                        for (String value: EntryUtils.toJSON(entry)) callback.processEntry(key, value);
                    } catch (Exception e) {
                        logger.error("CanalSource 处理 Entiry 失败", e);
                    }
                }
            }

            canalConnector.ack(messageId);
        } catch (CanalClientException e) {
            logger.error("CanalClient 异常，重启试试", e);
            this.destroy();
            this.init();
        } catch (Exception e) {
            logger.error("CanalSource 处理信息失败，数据回滚", e);
            if (messageId != -1) canalConnector.rollback(messageId);
        }
    }

    @Override
    public void destroy() {
        this.canalConnector.disconnect();
        this.canalConnector = null;
    }
}

File: src/main/scala/com/dyingbleed/akita/stream/CanalSource.scala
package com.dyingbleed.akita.stream

import java.net.InetSocketAddress
import java.util

import akka.stream.stage.{GraphStage, GraphStageLogic, OutHandler, StageLogging}
import akka.stream.{Attributes, Outlet, SourceShape}
import com.alibaba.otter.canal.client.{CanalConnector, CanalConnectors}
import com.alibaba.otter.canal.protocol.CanalEntry.EntryType
import com.dyingbleed.akita.utils.EntryUtils
import com.google.common.collect.Lists
import com.google.common.net.HostAndPort
import org.apache.commons.lang.StringUtils

import scala.collection.JavaConverters._
import scala.collection.mutable

/**
  * Canal Source
  *
  * @param servers  Canal 服务地址
  * @param destination  Canal 服务部署的实例
  * @param username  用户名
  * @param password  密码
  * @param filter  schema 过滤
  *
  * Created by 李震 on 2018/5/5.
  */
class CanalSource(
                   servers: String,
                   destination: String,
                   username: String,
                   password: String,
                   filter: String
                 ) extends GraphStage[SourceShape[String]] {

  val out: Outlet[String] = Outlet("canal.out")

  override def shape: SourceShape[String] = SourceShape(out)

  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
    new GraphStageLogic(shape) with StageLogging {

      private var canalConnector: CanalConnector = _

      private val messageQueue = new mutable.Queue[String]()

      private var messageBatchId = -1l

      override def preStart(): Unit = {
        if (this.canalConnector == null) {
          this.canalConnector = if (servers.contains(",")) {
            // 集群模式
            val hosts: util.LinkedList[InetSocketAddress] = Lists.newLinkedList()
            for (server <- StringUtils.split(servers, ",")) {
              val hostAndPort = HostAndPort.fromString(server)
              hosts.add(new InetSocketAddress(hostAndPort.getHostText, hostAndPort.getPortOrDefault(11111)))
            }
            CanalConnectors.newClusterConnector(hosts, destination, username, password)
          } else {
            // 单机模式
            val hostAndPort: HostAndPort = HostAndPort.fromString(servers)
            CanalConnectors.newSingleConnector(
              new InetSocketAddress(hostAndPort.getHostText, hostAndPort.getPortOrDefault(11111)),
              destination,
              username,
              password
            )
          }

          log.info("连接 canal {}", servers)
          this.canalConnector.connect()
          this.canalConnector.subscribe(filter)
          this.canalConnector.rollback()
          log.info("连接 canal 成功")
        }
      }

      setHandler(out, new OutHandler {
        override def onPull(): Unit = {
          /*
           * 1. 如果队列为空，从 canal 获取批量数据（最多 100 条），入队
           * */
          while (messageQueue.isEmpty) {
            val messageBatch = canalConnector.getWithoutAck(100)
            messageBatchId = messageBatch.getId
            log.info("收到 canal 消息 {}", messageBatchId)
            if (messageBatchId != -1) {
              for (entry <- messageBatch.getEntries.asScala) {
                if (entry.getEntryType == EntryType.TRANSACTIONBEGIN || entry.getEntryType == EntryType.TRANSACTIONEND
                ) {
                  // do nothing
                } else {
                  for (json <- EntryUtils.toJSON(entry).asScala) {
                    messageQueue.enqueue(json)
                  }
                }
              }
            }
          }

          /*
           * 2. 出队，推送到出口
           * */
          val message = messageQueue.dequeue()
          push(out, message)

          /*
           * 3. 如果队列为空，即上次获取批量数据处理完毕，向 canal 确认
           * */
          if (messageQueue.isEmpty && messageBatchId != -1) canalConnector.ack(messageBatchId)
        }
      })

      override def postStop(): Unit = {
        this.canalConnector.disconnect()
        this.canalConnector = null
        log.info("canal 连接关闭")
      }
    }
  }

}
